Build a regression model.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import requests
import pandas as pd
import os
import json
stations_df = pd.read_csv('citybikes_stations.csv')
venues_df= pd.read_csv('correct_foursquare_venues.csv')
restaurants_df=pd.read_csv('correct_yelp_restaurants.csv')

merged_df_r = pd.merge(stations_df, restaurants_df, on=('Latitude','Longitude'))
merged_df_v = pd.merge(stations_df, venues_df, on=('Latitude','Longitude'))


Mean Squared Error: 0.7154265436231191
R-squared: 0.9959194645539086


In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


# Select features and target variable
features = ['Bikes Available', 'Slots','Ebikes']
target = 'Empty Slots'


# Prepare features (X) and target variable (y)
X = merged_df_r[features]
y = merged_df_r[target]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the linear regression model
model = LinearRegression()

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
predictions = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, predictions)
r2 = r2_score(y_test, predictions)

print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')
print(f'intercept:{model.intercept_}')
print(f'coefficients:{model.coef_}')

Mean Squared Error: 0.7154265436231191
R-squared: 0.9959194645539086
intercept:-0.6325863295735772
coefficients:[-0.94975819  1.00883898 -0.28349051]


Provide model output and an interpretation of the results. 

In [ ]:
Mean Squared Error: 0.7988213475482295
R-squared: 0.9957857549311925


1. **Intercept**: The intercept, represented as -0.6325863295735772, is the value of the dependent variable when all independent variables are zero. It indicates the baseline value of the dependent variable in the absence of any predictor variables.

2. **Coefficients**: The coefficients represent the effect of each independent variable on the dependent variable, holding other variables constant. In your model:
   - The coefficient of the first independent variable is approximately -0.94975819.
   - The coefficient of the second independent variable is approximately 1.00883898.
   - The coefficient of the third independent variable is approximately -0.28349051.
   
   These values indicate the change in the dependent variable for a one-unit change in the respective independent variable, while keeping other variables constant. A negative coefficient suggests a negative correlation, and a positive coefficient suggests a positive correlation.

3. **Mean Squared Error (MSE)**: A lower MSE indicates that the model's predictions are closer to the actual values on average.

4. **R-squared (R²)**:  an R-squared value of approximately 0.9958 means that about 99.59% of the variance in the dependent variable is predictable from the independent variables in your model. A higher R-squared value indicates a better fit of the model to the data.



# Stretch

How can you turn the regression model into a classification model?

In [6]:
low_threshold = 10  # Adjust this value based on your dataset
medium_threshold = 30  # Adjust this value based on your dataset

# Define the bins and labels for classes
bins = [-float("inf"), low_threshold, medium_threshold, float("inf")]
labels = ['Unlikely', 'Likely', 'Highly']

# Binning the 'Empty Slots' into classes
merged_df_r['Empty Slots Class'] = pd.cut(merged_df_r['Empty Slots'], bins=bins, labels=labels)

features = ['Rating', 'Reviews','Banking','Slots']
target = 'Empty Slots Class'  # The new target variable

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Prepare features (X) and target variable (y)
X = merged_df_r[features]
y = merged_df_r[target]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Decision Tree classifier
classifier = DecisionTreeClassifier(random_state=42)

# Train the classifier on the training data
classifier.fit(X_train, y_train)

# Make predictions on the test data
predictions = classifier.predict(X_test)

# Evaluate the classifier
accuracy = accuracy_score(y_test, predictions)
print(f'Accuracy: {accuracy}')




Accuracy: 0.9781659388646288


DecisionTreeClassifier Model